pkg/controller/bundles/bundles_syncer.go

```diff
@@ -3,118 +3,59 @@ package bundles
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-const (
-	envVarK8sClientsPoolSize  = "K8S_CLIENTS_POOL_SIZE"
-	defaultK8sClientsPoolSize = 10
-)
-
-// LeafHubBundlesSpecSync syncs bundles spec objects.
-type LeafHubBundlesSpecSync struct {
-	log                    logr.Logger
-	bundleUpdatesChan      chan *bundle.ObjectsBundle
-	k8sClientsPool         []client.Client
-	clientWorkersJobChan   chan *clientWorkerJob
-	clientWorkersWaitGroup sync.WaitGroup
-}
-
-// handlerFunc is a clientWorkerJob's handler function.
-type handlerFunc func(context.Context, client.Client, *unstructured.Unstructured)
-
-// clientWorkerJob holds the object than need to be processed and the flag to which defines
-// whether object need to be updated or delete.
-type clientWorkerJob struct {
-	handler handlerFunc
-	obj     *unstructured.Unstructured
-}
-
 // AddLeafHubBundlesSpecSync adds bundles spec syncer to the manager.
 func AddLeafHubBundlesSpecSync(log logr.Logger, mgr ctrl.Manager, bundleUpdatesChan chan *bundle.ObjectsBundle) error {
-	k8sClientsPoolSize := readK8sPoolSizeEnvVar(log)
-
-	// creates the in-cluster config
-	config, err := rest.InClusterConfig()
+	// create k8s worker pool
+	k8sWorkerPool, err := k8sworkerpool.NewK8sWorkerPool(log)
 	if err != nil {
-		return fmt.Errorf("failed to get in cluster kubeconfig - %w", err)
+		return fmt.Errorf("failed to initialize bundle spec syncer - %w", err)
 	}
 
-	// prepare k8s clients pool
-	k8sClientsPool := make([]client.Client, k8sClientsPoolSize)
-
-	for i := 0; i < k8sClientsPoolSize; i++ {
-		k8sClient, err := client.New(config, client.Options{})
-		if err != nil {
-			return fmt.Errorf("failed to initialize k8s client - %w", err)
-		}
-
-		k8sClientsPool[i] = k8sClient
-	}
-
-	// create client workers job channel
-	clientWorkersJobChan := make(chan *clientWorkerJob, k8sClientsPoolSize)
-
 	if err := mgr.Add(&LeafHubBundlesSpecSync{
-		log:                  log,
-		bundleUpdatesChan:    bundleUpdatesChan,
-		k8sClientsPool:       k8sClientsPool,
-		clientWorkersJobChan: clientWorkersJobChan,
+		log:                          log,
+		bundleUpdatesChan:            bundleUpdatesChan,
+		k8sWorkerPool:                k8sWorkerPool,
+		bundleProcessingWaitingGroup: sync.WaitGroup{},
 	}); err != nil {
 		return fmt.Errorf("failed to add bundles spec syncer - %w", err)
 	}
 
 	return nil
 }
 
+// LeafHubBundlesSpecSync syncs bundles spec objects.
+type LeafHubBundlesSpecSync struct {
+	log                          logr.Logger
+	bundleUpdatesChan            chan *bundle.ObjectsBundle
+	k8sWorkerPool                *k8sworkerpool.K8sWorkerPool
+	bundleProcessingWaitingGroup sync.WaitGroup
+}
+
 // Start function starts bundles spec syncer.
 func (syncer *LeafHubBundlesSpecSync) Start(stopChannel <-chan struct{}) error {
 	ctx, cancelContext := context.WithCancel(context.Background())
 	defer cancelContext()
 
-	// start workers
-	for i := 0; i < len(syncer.k8sClientsPool); i++ {
```

Comment:
I don't see you call syncer.k8sWorkerPool.Start()

Reply:
good catch. fixed.